File: ios/AnalysisOverlayRef.swift
import ArcGIS
import ExpoModulesCore

/// SharedObject wrapping a native `AnalysisOverlay` owned by a SceneView. Holds visual analyses
/// (`<Viewshed>` / `<LineOfSight>`) declared as children of `<AnalysisOverlay>`.
public class AnalysisOverlayRef: SharedObject {
  let overlay = AnalysisOverlay()

  func addAnalysis(_ ref: AnalysisRef) {
    overlay.addAnalysis(ref.analysis)
  }

  func removeAnalysis(_ ref: AnalysisRef) {
    overlay.removeAnalysis(ref.analysis)
  }

  func setVisible(_ visible: Bool) {
    overlay.isVisible = visible
  }
}

/// Base SharedObject wrapping an exploratory `Analysis` (viewshed / line-of-sight); the overlay
/// reads `analysis` by reference. Subclasses build and own the concrete analysis.
public class AnalysisRef: SharedObject {
  let analysis: Analysis

  init(analysis: Analysis) {
    self.analysis = analysis
    super.init()
  }
}

/// SharedObject wrapping an `ExploratoryLocationViewshed` — the area visible from an observer.
public final class ViewshedRef: AnalysisRef {
  private let viewshed: ExploratoryLocationViewshed

  init(props: [String: Any]) {
    let location = (props["location"] as? [String: Any]).flatMap(geometryFromDict) as? Point
      ?? Point(x: 0, y: 0, spatialReference: .wgs84)
    let viewshed = ExploratoryLocationViewshed(
      location: location,
      heading: (props["heading"] as? NSNumber)?.doubleValue ?? 0,
      pitch: (props["pitch"] as? NSNumber)?.doubleValue ?? 90,
      horizontalAngle: (props["horizontalAngle"] as? NSNumber)?.doubleValue ?? 45,
      verticalAngle: (props["verticalAngle"] as? NSNumber)?.doubleValue ?? 45,
      minDistance: (props["minDistance"] as? NSNumber)?.doubleValue,
      maxDistance: (props["maxDistance"] as? NSNumber)?.doubleValue
    )
    self.viewshed = viewshed
    super.init(analysis: viewshed)
    applyProps(props)
  }

  func applyProps(_ changed: [String: Any]) {
    for (key, value) in changed {
      switch key {
      case "location":
        if let point = (value as? [String: Any]).flatMap(geometryFromDict) as? Point {
          viewshed.location = point
        }
      case "heading":
        if let n = value as? NSNumber { viewshed.heading = n.doubleValue }
      case "pitch":
        if let n = value as? NSNumber { viewshed.pitch = n.doubleValue }
      case "horizontalAngle":
        if let n = value as? NSNumber { viewshed.horizontalAngle = n.doubleValue }
      case "verticalAngle":
        if let n = value as? NSNumber { viewshed.verticalAngle = n.doubleValue }
      case "minDistance":
        viewshed.minDistance = (value as? NSNumber)?.doubleValue
      case "maxDistance":
        viewshed.maxDistance = (value as? NSNumber)?.doubleValue
      case "frustumOutlineVisible":
        if let b = value as? Bool { viewshed.frustumOutlineIsVisible = b }
      default:
        break
      }
    }
  }
}

/// SharedObject wrapping an `ExploratoryGeoElementViewshed` — a viewshed whose observer
/// tracks a `Graphic` (GeoElement) as it moves. Constructed with a `GraphicRef` and view props.
public final class GeoElementViewshedRef: AnalysisRef {
  private let viewshed: ExploratoryGeoElementViewshed

  init(graphic: GraphicRef, props: [String: Any]) {
    let viewshed = ExploratoryGeoElementViewshed(
      geoElement: graphic.graphic,
      horizontalAngle: (props["horizontalAngle"] as? NSNumber)?.doubleValue ?? 45,
      verticalAngle: (props["verticalAngle"] as? NSNumber)?.doubleValue ?? 45,
      headingOffset: (props["headingOffset"] as? NSNumber)?.doubleValue ?? 0,
      pitchOffset: (props["pitchOffset"] as? NSNumber)?.doubleValue ?? 0,
      minDistance: (props["minDistance"] as? NSNumber)?.doubleValue,
      maxDistance: (props["maxDistance"] as? NSNumber)?.doubleValue
    )
    self.viewshed = viewshed
    super.init(analysis: viewshed)
    applyProps(props)
  }

  func applyProps(_ changed: [String: Any]) {
    for (key, value) in changed {
      switch key {
      case "headingOffset":
        if let n = value as? NSNumber { viewshed.headingOffset = n.doubleValue }
      case "pitchOffset":
        if let n = value as? NSNumber { viewshed.pitchOffset = n.doubleValue }
      case "horizontalAngle":
        if let n = value as? NSNumber { viewshed.horizontalAngle = n.doubleValue }
      case "verticalAngle":
        if let n = value as? NSNumber { viewshed.verticalAngle = n.doubleValue }
      case "minDistance":
        viewshed.minDistance = (value as? NSNumber)?.doubleValue
      case "maxDistance":
        viewshed.maxDistance = (value as? NSNumber)?.doubleValue
      case "frustumOutlineVisible":
        if let b = value as? Bool { viewshed.frustumOutlineIsVisible = b }
      default:
        break
      }
    }
  }
}

/// SharedObject wrapping an `ExploratoryLocationLineOfSight` — observer→target visibility.
/// Streams the target's visibility back to JS via `onTargetVisibilityChange`.
public final class LineOfSightRef: AnalysisRef {
  private let lineOfSight: ExploratoryLocationLineOfSight
  private var observation: Task<Void, Never>?

  init(props: [String: Any]) {
    let observer = (props["observer"] as? [String: Any]).flatMap(geometryFromDict) as? Point
      ?? Point(x: 0, y: 0, spatialReference: .wgs84)
    let target = (props["target"] as? [String: Any]).flatMap(geometryFromDict) as? Point
      ?? Point(x: 0, y: 0, spatialReference: .wgs84)
    let lineOfSight = ExploratoryLocationLineOfSight(observerLocation: observer, targetLocation: target)
    self.lineOfSight = lineOfSight
    super.init(analysis: lineOfSight)
    observation = Task { [weak self] in
      guard let stream = self?.lineOfSight.$targetVisibility else { return }
      for await visibility in stream {
        guard let self else { break }
        self.emit(event: "onTargetVisibilityChange", payload: ["visibility": visibilityString(visibility)])
      }
    }
  }

  override public func sharedObjectWillRelease() {
    observation?.cancel()
    observation = nil
    super.sharedObjectWillRelease()
  }

  func applyProps(_ changed: [String: Any]) {
    for (key, value) in changed {
      switch key {
      case "observer":
        if let point = (value as? [String: Any]).flatMap(geometryFromDict) as? Point {
          lineOfSight.observerLocation = point
        }
      case "target":
        if let point = (value as? [String: Any]).flatMap(geometryFromDict) as? Point {
          lineOfSight.targetLocation = point
        }
      default:
        break
      }
    }
  }
}

/// SharedObject wrapping an `ExploratoryLocationDistanceMeasurement` — direct/horizontal/vertical
/// distance between two 3D points. Streams the measurements to JS via `onMeasurementChange`.
public final class DistanceMeasurementRef: AnalysisRef {
  private let measurement: ExploratoryLocationDistanceMeasurement
  private var observation: Task<Void, Never>?

  init(props: [String: Any]) {
    let start = (props["startLocation"] as? [String: Any]).flatMap(geometryFromDict) as? Point
      ?? Point(x: 0, y: 0, spatialReference: .wgs84)
    let end = (props["endLocation"] as? [String: Any]).flatMap(geometryFromDict) as? Point
      ?? Point(x: 0, y: 0, spatialReference: .wgs84)
    let measurement = ExploratoryLocationDistanceMeasurement(startLocation: start, endLocation: end)
    self.measurement = measurement
    super.init(analysis: measurement)
    observation = Task { [weak self] in
      guard let stream = self?.measurement.measurements else { return }
      for await m in stream {
        guard let self else { break }
        self.emit(event: "onMeasurementChange", payload: [
          "directDistance": m.directDistance.value,
          "horizontalDistance": m.horizontalDistance.value,
          "verticalDistance": m.verticalDistance.value,
        ])
      }
    }
  }

  override public func sharedObjectWillRelease() {
    observation?.cancel()
    observation = nil
    super.sharedObjectWillRelease()
  }

  func applyProps(_ changed: [String: Any]) {
    for (key, value) in changed {
      switch key {
      case "startLocation":
        if let p = (value as? [String: Any]).flatMap(geometryFromDict) as? Point { measurement.startLocation = p }
      case "endLocation":
        if let p = (value as? [String: Any]).flatMap(geometryFromDict) as? Point { measurement.endLocation = p }
      default:
        break
      }
    }
  }
}

/// SharedObject wrapping an `ExploratoryGeoElementLineOfSight` — a line of sight whose observer
/// and target both track `Graphic`s (GeoElements) as they move.
/// Streams the target's visibility back to JS via `onTargetVisibilityChange`.
public final class GeoElementLineOfSightRef: AnalysisRef {
  private let lineOfSight: ExploratoryGeoElementLineOfSight
  private var observation: Task<Void, Never>?

  init(observer: GraphicRef, target: GraphicRef) {
    let lineOfSight = ExploratoryGeoElementLineOfSight(
      observer: observer.graphic,
      target: target.graphic
    )
    self.lineOfSight = lineOfSight
    super.init(analysis: lineOfSight)
    observation = Task { [weak self] in
      guard let stream = self?.lineOfSight.$targetVisibility else { return }
      for await visibility in stream {
        guard let self else { break }
        self.emit(event: "onTargetVisibilityChange", payload: ["visibility": visibilityString(visibility)])
      }
    }
  }

  override public func sharedObjectWillRelease() {
    observation?.cancel()
    observation = nil
    super.sharedObjectWillRelease()
  }
}

/// Maps the native line-of-sight visibility enum to the JS `TargetVisibility` union.
private func visibilityString(_ v: ExploratoryLineOfSight.TargetVisibility) -> String {
  switch v {
  case .visible: return "visible"
  case .obstructed: return "obstructed"
  case .unknown: return "unknown"
  @unknown default: return "unknown"
  }
}

File: android/src/main/java/expo/modules/arcgis/AnalysisOverlayRef.kt
package expo.modules.arcgis

import com.arcgismaps.analysis.interactive.Analysis
import com.arcgismaps.analysis.interactive.ExploratoryLineOfSightTargetVisibility
import com.arcgismaps.analysis.interactive.ExploratoryLocationDistanceMeasurement
import com.arcgismaps.analysis.interactive.ExploratoryLocationLineOfSight
import com.arcgismaps.analysis.interactive.ExploratoryLocationViewshed
import com.arcgismaps.geometry.Point
import com.arcgismaps.geometry.SpatialReference
import com.arcgismaps.mapping.view.AnalysisOverlay
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.sharedobjects.SharedObject
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * SharedObject wrapping a native [AnalysisOverlay] owned by a SceneView. Holds the visual analyses
 * (`<Viewshed>` / `<LineOfSight>`) declared as children of `<AnalysisOverlay>`.
 */
class AnalysisOverlayRef(appContext: AppContext) : SharedObject(appContext) {
  val overlay = AnalysisOverlay(emptyList())

  fun addAnalysis(ref: AnalysisRef) {
    overlay.analyses.add(ref.analysis)
  }

  fun removeAnalysis(ref: AnalysisRef) {
    overlay.analyses.remove(ref.analysis)
  }

  fun setVisible(visible: Boolean) {
    overlay.isVisible = visible
  }
}

/** Base SharedObject wrapping an exploratory [Analysis] (viewshed / line-of-sight). */
abstract class AnalysisRef(appContext: AppContext) : SharedObject(appContext) {
  abstract val analysis: Analysis
}

/** SharedObject wrapping an [ExploratoryLocationViewshed] — the area visible from an observer. */
class ViewshedRef(appContext: AppContext, props: Map<String, Any?>) : AnalysisRef(appContext) {
  private val viewshed = ExploratoryLocationViewshed(
    analysisPoint(props["location"]) ?: Point(0.0, 0.0, SpatialReference.wgs84()),
    numOr(props["heading"], 0.0),
    numOr(props["pitch"], 90.0),
    numOr(props["horizontalAngle"], 45.0),
    numOr(props["verticalAngle"], 45.0),
    (props["minDistance"] as? Number)?.toDouble(),
    (props["maxDistance"] as? Number)?.toDouble(),
  )

  override val analysis: Analysis get() = viewshed

  init {
    applyProps(props)
  }

  fun applyProps(changed: Map<String, Any?>) {
    changed.forEach { (key, value) ->
      when (key) {
        "location" -> analysisPoint(value)?.let { viewshed.location = it }
        "heading" -> (value as? Number)?.toDouble()?.let { viewshed.heading = it }
        "pitch" -> (value as? Number)?.toDouble()?.let { viewshed.pitch = it }
        "horizontalAngle" -> (value as? Number)?.toDouble()?.let { viewshed.horizontalAngle = it }
        "verticalAngle" -> (value as? Number)?.toDouble()?.let { viewshed.verticalAngle = it }
        "minDistance" -> viewshed.minDistance = (value as? Number)?.toDouble()
        "maxDistance" -> viewshed.maxDistance = (value as? Number)?.toDouble()
        "frustumOutlineVisible" -> (value as? Boolean)?.let { viewshed.frustumOutlineVisible = it }
      }
    }
  }
}

/** SharedObject wrapping an [ExploratoryLocationLineOfSight] — streams target visibility to JS. */
class LineOfSightRef(appContext: AppContext, props: Map<String, Any?>) : AnalysisRef(appContext) {
  private val lineOfSight = ExploratoryLocationLineOfSight(
    analysisPoint(props["observer"]) ?: Point(0.0, 0.0, SpatialReference.wgs84()),
    analysisPoint(props["target"]) ?: Point(0.0, 0.0, SpatialReference.wgs84()),
  )
  private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

  override val analysis: Analysis get() = lineOfSight

  init {
    scope.launch {
      lineOfSight.targetVisibility.collect { visibility ->
        emit("onTargetVisibilityChange", mapOf("visibility" to visibilityString(visibility)))
      }
    }
  }

  override fun deallocate() {
    scope.cancel()
    super.deallocate()
  }

  fun applyProps(changed: Map<String, Any?>) {
    changed.forEach { (key, value) ->
      when (key) {
        "observer" -> analysisPoint(value)?.let { lineOfSight.observerLocation = it }
        "target" -> analysisPoint(value)?.let { lineOfSight.targetLocation = it }
      }
    }
  }
}

/**
 * SharedObject wrapping an [ExploratoryLocationDistanceMeasurement] — direct/horizontal/vertical
 * distance between two 3D points. Streams the measurements to JS via `onMeasurementChange`.
 */
class DistanceMeasurementRef(appContext: AppContext, props: Map<String, Any?>) : AnalysisRef(appContext) {
  private val measurement = ExploratoryLocationDistanceMeasurement(
    analysisPoint(props["startLocation"]) ?: Point(0.0, 0.0, SpatialReference.wgs84()),
    analysisPoint(props["endLocation"]) ?: Point(0.0, 0.0, SpatialReference.wgs84()),
  )
  private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

  override val analysis: Analysis get() = measurement

  init {
    scope.launch {
      measurement.measurementChanged.collect { m ->
        emit(
          "onMeasurementChange",
          mapOf(
            "directDistance" to m.directDistance.value,
            "horizontalDistance" to m.horizontalDistance.value,
            "verticalDistance" to m.verticalDistance.value,
          ),
        )
      }
    }
  }

  override fun deallocate() {
    scope.cancel()
    super.deallocate()
  }

  fun applyProps(changed: Map<String, Any?>) {
    changed.forEach { (key, value) ->
      when (key) {
        "startLocation" -> analysisPoint(value)?.let { measurement.startLocation = it }
        "endLocation" -> analysisPoint(value)?.let { measurement.endLocation = it }
      }
    }
  }
}

/** Decodes a JS point dict into a [Point] (returns null if the value is not a point geometry). */
internal fun analysisPoint(value: Any?): Point? =
  (value as? Map<*, *>)?.let { geometryFromDict(it) } as? Point

private fun numOr(value: Any?, default: Double): Double = (value as? Number)?.toDouble() ?: default

/** Maps the native line-of-sight visibility to the JS `TargetVisibility` union. */
private fun visibilityString(v: ExploratoryLineOfSightTargetVisibility): String = when (v) {
  is ExploratoryLineOfSightTargetVisibility.Visible -> "visible"
  is ExploratoryLineOfSightTargetVisibility.Obstructed -> "obstructed"
  else -> "unknown"
}
